### PR TITLE
Use estimated realtime arrivals at stops

### PR DESCRIPTION
**Pull request by scascketta**
This uses the CapMetro endpoint at `/planner/s_nextbus2.asp` which provides estimated realtime arrival info
Note that this is the same data used in the official CapMetro apps.

### CODE REVIEW

File: src/js/models/Stop.js [hunk 1]
```diff
@@ -72,9 +72,9 @@ Stop.prototype = {
                 this.errorMsg(msg);
             }.bind(this))
             .then(function() {
-                var activeTrip = this.stopDetails.tripCollections()[0];
-                if (activeTrip.trips()[0]) {
-                    this.trips(activeTrip.trips());
+                var tc = this.stopDetails.tripCollection;
```
Comment (luqmaan) [inline src/js/models/Stop.js]: Why not just use `this.stopDetails.tripCollection`? Seems more clear that way.

Comment (scascketta) [inline src/js/models/Stop.js]: Foolishly saving characters is all. I think it's fine as:

```
this.trips(this.stopDetails.tripCollection.trips());
```

File: src/js/models/StopDetails.js [hunk 2]
```diff
@@ -40,12 +34,8 @@ StopDetails.prototype.fetch = function() {
...
         requests.get(yqlURL, params)
             .then(this.parseResponse.bind(this))
-            .tap(function(Services) {
-                var tripCollections = Services.map(function(Service) {
-                    return new TripCollection(this.stopID(), Service);
-                }.bind(this));
-
-                this.tripCollections(tripCollections);
+            .tap(function(Runs) {
+                this.tripCollection = new TripCollection(this.stopID(), Runs);
```
Comment (luqmaan) [inline src/js/models/StopDetails.js]: This isn't directly related to this PR, but I think we could simplify retryAtMost by moving the catch stuff into their own functions.

```js
    function retryAtMost(maxRetries) {
        requests.get(yqlURL, params)
            .then(this.parseResponse.bind(this))
            .tap(function(Runs) {
                this.tripCollection = new TripCollection(this.stopID(), Runs);
                deferred.resolve();
            }.bind(this))
            .catch(CapMetroAPIError, handleCapMetroAPIError.bind(this, maxRetries))
            .catch(function(err) {
                this.errorMsg(err);
                deferred.reject(err);
            }.bind(this))
            .finally(function() {
                this.loading(false);
            }.bind(this));
    }

    function handleCapMetroAPIError(maxRetries, err) {
        var msg = err.message + '. Retrying ' + maxRetries + ' more times';
        console.error(msg);
        this.errorMsg(msg);

        if (maxRetries > 0) {
            return retryAtMost.call(this, maxRetries - 1);
        }
        else {
            deferred.reject(CapMetroAPIError);
        }
    }

    retryAtMost.call(this, config.MAX_RETRIES);

    return deferred.promise;
};
```

Generally speaking, this retry stuff could use some rethinking.

File: src/js/models/StopDetails.js [hunk 3]
```diff
@@ -89,13 +80,18 @@ StopDetails.prototype.parseResponse = function(res) {
 
         throw new Error(faultcode + ' ' + faultstring);
     }
+    Runs = res.query.results.Envelope.Body.Nextbus2Response.Runs.Run;
```
Comment (luqmaan) [inline src/js/models/StopDetails.js]: Sometimes Runs.Run will be a single `<Run>` element instead of a list.

Trapeze/ANUS has done this with `Trips` and `Vehicles`. See https://github.com/luqmaan/MetroRappid/blob/d418e4b422cb9e0e68e9251f1ae8935d1bc52ab3/src/js/models/VehicleCollection.js#L78-L82 and https://github.com/luqmaan/MetroRappid/blob/65e73177d341bc70a1676805733f9a7c335f3c5b/src/js/models/TripCollection.js#L18-L21

File: src/js/models/Trip.js [hunk 4]
```diff
@@ -2,9 +2,8 @@ var ko = require('knockout');
 var moment = require('moment');
 
 function Trip(data) {
-    this.tripTime = ko.observable(data.Triptime);
+    this.tripTime = ko.observable(data.Estimatedtime);
```
Comment (luqmaan) [inline src/js/models/Trip.js]: :sexy:

File: src/js/models/TripCollection.js [hunk 5]
```diff
@@ -5,23 +5,17 @@ var config = require('../config');
 var Trip = require('./Trip');
 
 
-function TripCollection(stopID, Service) {
+function TripCollection(stopID, Runs) {
     this.stopID = ko.observable(stopID);
+    this.routeID = ko.observable(Runs[0].Route);
```
Comment (luqmaan) [inline src/js/models/TripCollection.js]: Seems weird to `Runs[0]` here.

File: src/js/models/Stop.js (hunk 1 above)
Comment (luqmaan) [inline src/js/models/Stop.js]: :smirk_cat:

**Comment by luqmaan** [comment]
Can you check in a JSON-ified `NextBus2` response into `mockdata.json`?

File: src/js/models/StopDetails.js (hunk 3 above)
Comment (scascketta) [inline src/js/models/StopDetails.js]: I remember seeing that snippet around. While I never encountered it in testing, it's better to have it and not break it for now.

**Comment by luqmaan** [comment]
#174

**Comment by luqmaan** [comment]
Stuff works great, and the realtime estimates are much better!

File: src/js/models/TripCollection.js (hunk 5 above)
Comment (scascketta) [inline src/js/models/TripCollection.js]: Agreed, I'm thinking of making the TripCollection constructor look like this: 

```
function TripCollection(stopID, route, sign, Runs) {
    ...
}
```

File: src/js/models/StopDetails.js (hunk 2 above)
Comment (scascketta) [inline src/js/models/StopDetails.js]: I agree, I'm thinking `retryAtMost` should be an extension of `requests`, with `maxRetries` as an option passed to `requests.*()` calls.

Comment (luqmaan) [inline src/js/models/StopDetails.js]: That makes sense. Sounds like it should be in another PR.

File: src/js/models/StopDetails.js [hunk 6]
```diff
@@ -40,12 +34,8 @@ StopDetails.prototype.fetch = function() {
...
         requests.get(yqlURL, params)
             .then(this.parseResponse.bind(this))
-            .tap(function(Services) {
-                var tripCollections = Services.map(function(Service) {
-                    return new TripCollection(this.stopID(), Service);
-                }.bind(this));
-
-                this.tripCollections(tripCollections);
+            .tap(function(Runs) {
+                this.tripCollection = new TripCollection(this.stopID(), this.routeID(), Runs[0].Sign, Runs);
```
Comment (luqmaan) [inline src/js/models/StopDetails.js]: Why don't we just remove the `sign` param from the TripCollection constructor and let the constructor do the `this.sign = Runs[0]`.

**Comment by luqmaan** [comment]
:dancer: :tada: :dancer: :fireworks: :dancer:

File: src/js/models/StopDetails.js (hunk 6 above)
Comment (scascketta) [inline src/js/models/StopDetails.js]: :+1:
